### DISCUSSION TITLE
the clearErrors function sets the isValid attribute to true when the form is invalid

### DISCUSSION THREAD

**Discussion by matilameiro** [react-hook-form/react-hook-form#7517] [General] [answered]
First of all, I apologize for the English, I am using google translate.

I have a field using the controller component that has a component inside
from the react-formated-number library. In the onChange the credit card number is valid against the api. If it is not correct I use setError and if it is correct and in case I previously had an error set, I use the clearErrors function which sets the isValid attribute to true for the entire form and this is not correct.

**Comment by bluebill1049** [comment]
I believe this is fixed in the latest version

**Comment by Moshyfawn** [comment, marked as answer]
What's your react-hook-form package version? Works fine with the latest 7.23.0 version.
Check out [this codesandbox](https://codesandbox.io/s/clear-individual-errors-example-qwoz5) to see if you do anything different. Try your version of the react-hook-form package with the codesandbox.

1. Submit the form once for the first field validation error to kick in
2. Enter the full credit card number. The card validation will fail 
3. Change any credit card string number. The validation will be successful and only credit card field error will be cleared

**Reply by bluebill1049** [reply]
thank you @Moshyfawn again for your help
<img width="917" alt="Screen Shot 2022-01-13 at 11 13 26 am" src="https://user-images.githubusercontent.com/10513364/149243112-ce7d664c-6a13-4608-aa06-c6d45ca8ea32.png">
https://github.com/react-hook-form/react-hook-form/commit/c1d77313534c440bef3e3490569f46cdd395ce72

**Reply by Moshyfawn** [reply]
Aw, thx :P Luv the lib! Hoping to get some time to look into the RHF codebase to maybe do PRs in the future. Have time for discussions only at this time :(

**Comment by matilameiro** [comment]
Many thanks to both of you for the quick response. With the version change it was solved and thank you very much for the example.

Now the problem that is happening to me is that the error is eliminated but the isValid attribute remains false even if there is no error and it is the attribute that I use to see if it can advance in the stepper or not. I hope you can help me.

Thank you very much. Again I apologize for the english.

**Reply by matilameiro** [reply]
I forgot to mention that the validation is done in the onChange. In case it helps.

**Reply by Moshyfawn** [reply]
I can't seem to reproduce. The `isValid` flag is set to `true` is soon as the card number validation is successful. See this video:

https://user-images.githubusercontent.com/16290753/149337685-86b08f84-6bc2-48f0-a4e2-afd4766e36a7.mp4

Can you, maybe, edit [this codesanbox](https://codesandbox.io/s/clear-errors-form-valid-fiv3y) to closer match what you're experiencing?

**Reply by matilameiro** [reply]
I reproduced roughly what happens to me. The code is not quite right. Take out some stuff and reuse the validation function but the same thing happens with the isValid attribute.
Here is the link https://codesandbox.io/s/clear-errors-form-valid-forked-j7lpu?file=/src/App.tsx

**Reply by matilameiro** [reply]
![error_isValid_1](https://user-images.githubusercontent.com/18328969/149352661-83f2b88a-f3e4-427e-9a55-10445537d5f0.PNG)
![error_isValid_2](https://user-images.githubusercontent.com/18328969/149352666-894751e4-6f06-48b1-8c6a-729f2fd5c88a.PNG)

This is my code in case it is helpful

**Reply by Moshyfawn** [reply]
OK, I think I've manage to localize the problem. The short answer is you want to set [`mode` prop on `useForm`](https://react-hook-form.com/api/useform) to `'onChange'`. That's because the default `mode` is `'onSubmit'`, so you're expected to submit the form before you get the "re-validation on input value change" behavior (see [`reValidateMode` prop on `useForm`](https://react-hook-form.com/api/useform)).
Here's an example of your form working as you've expected: [codesandbox](https://codesandbox.io/s/clear-errors-form-valid-forked-l88wl?file=/src/App.tsx)

**Reply by Moshyfawn** [reply]
There's something about this behavior that feels incorrect. @bluebill1049 is it expected?
If the form `mode` is `'onSubmit'` and I kick in my own **custom validation** via `setError` **without** ever **submitting** the form, is `clearErrors` expected to **not** reset the `isValid` from state?

**Reply by Moshyfawn** [reply]
See #7524

**Reply by matilameiro** [reply]
The mode of the form is 'onBlur'.

Another problem I am having is that when I do onBlur of the second field (expiration, in the picture) it changes the isValid atributte to true when the credit card field has an error.

![error_isValid_3](https://user-images.githubusercontent.com/18328969/149386160-7753f63f-b188-4c2c-a7ae-6a6bee5773de.PNG)

**Reply by bluebill1049** [reply]
thanks @Moshyfawn I have replied to the issue. 

`clearErrors` should not affect isValid `formState`, because it doesn't change any validation rules, it only wipes `errors`, isValid should only reflect your form is valid or not.

setting a custom error that doesn't associate with input will require a manual `clearError` because it doesn't associate with any input.

**Reply by Moshyfawn** [reply]
Got it, @bluebill1049!
So, @matilameiro, you can potentially move your validation logic into[ the `validate` callback in `rules`](https://react-hook-form.com/api/usecontroller/controller). This will associate the potential card number error with the card number field. Since the form validation mode is `'onBlur'`, the credit card field is going to be validated on it's blur event. See [this codesandbox](https://codesandbox.io/s/custom-vs-register-form-validation-rms46).
I'm not sure if you can achieve the `onChange` validation for the credit card field without switching the form `mode` to `"onChange"`.

https://user-images.githubusercontent.com/16290753/149419090-7f962476-cfc5-453d-9a42-942621de5c2b.mp4

**Reply by matilameiro** [reply]
I tried to pass the validation to the validate but it happened more times than only in the onblur and it makes a lot of calls to the backend. In the onblur of the rest of the changes it also passes through the validate. I suspect that the combination with the react-format-number library is the big problem.

The other big problem is that the onBlur of the expiration field (which also has the react-format-number component) changes the isValid attribute. In my case it leaves it true.

I am thinking of changing one of the 2 libraries, no matter how long it takes me.

I am grateful for the responses. I thought I was not going to have such quick answers.
